bugsnag/bugsnagger.go

```diff
@@ -0,0 +1,262 @@
+// It formats the rawData to the general MetaData format for the bugsnag-go API.
+// Bugsnagger also contains the implementation for the external APIs that are defined in bugsnag.go.
+type bugsnagger struct {
+	config   *bugsnaggo.Configuration
+	notifier notifier
+}
+
+func NewBugsnagger(config *bugsnaggo.Configuration, notifier notifier) *bugsnagger {
+	return &bugsnagger{config: config, notifier: notifier}
+}
+
+// Optional metadata can be passed in as well, however only certain types of those will have an effect:
+// * string will be used to set the BS context (usually url path or goroutine identifier)
+// * http.Request creates Request tab in BS and tries to set context
+// * log.Entry/log.Fields creates Details tab reflecting the entry fields
+// * Tab allows creating arbitrary tab
+func (snagger *bugsnagger) buildData(err error, rawData ...interface{}) ([]interface{}, error) {
+	var (
+		user    bugsnaggo.User
+		req     *http.Request
+		context string
+		cause   error
+	)
+
+	bugsnagData := []interface{}{bugsnaggo.SeverityError}
+	md := make(bugsnaggo.MetaData)
+
+	if err != nil {
+		bugsnagData = append(bugsnagData, bugsnaggo.ErrorClass{Name: extractErrorClass(err)})
+		context, cause = snagger.formatError(err)
+		//set the Error tab with details/stacktrace
+		md["Error"] = Rows{"details": fmt.Sprintf("%+v", err)}
+	}
+
+	//extract more context/metadata from the rawData
+	for _, v := range rawData {
+		switch v := v.(type) {
+		case string:
+			context = v
+		case *http.Request:
+			req = v
+			bugsnagData = append(bugsnagData, req)
+		case *log.Entry:
+			userID, message, mdData := snagger.processLogEntry(v)
+			snagger.updateUserID(&user, userID)
+			if context == "" {
+				context = message
+			}
+			md["Log"] = mdData
+		case log.Fields:
+			userID := snagger.extractUserID(v)
+			snagger.updateUserID(&user, userID)
+			md["Log"] = v
+		case TabWriter:
+			tab := v.CreateBugsnagTab()
+			md[tab.Label] = tab.Rows
+		case Tab:
+			md[v.Label] = v.Rows
+		default:
+			bugsnagData = append(bugsnagData, v)
+		}
+	}
+
+	if len(md) > 0 {
+		bugsnagData = append(bugsnagData, md)
+	}
+	if user.Id != "" {
+		bugsnagData = append(bugsnagData, user)
+	}
+	if context == "" && req != nil && req.URL != nil {
+		context = req.URL.Path
+	}
+	bugsnagData = append(bugsnagData, bugsnaggo.Context{String: context})
+
+	return bugsnagData, cause
+}
+
+func (snagger *bugsnagger) formatError(err error) (string, error) {
+	var context string
+	cause := errors.Cause(err)
+	if uerr, ok := cause.(*url.Error); ok {
+		cause = uerr.Err
+	} else if serr, ok := cause.(*http2.StreamError); ok {
+		if serr.Cause != nil {
+			cause = serr.Cause
+		} else {
+			cause = fmt.Errorf("stream error: %s", serr.Code)
+		}
+	}
+
+	// Try to get the wrapped message to set the bugsnag context
+	errString := fmt.Sprintf("%s", err)
+	// Note that error.Wrap always appends a space after the :
+	i := strings.LastIndex(errString, ": ")
```

Comment:
could this just be
```
context = strings.TrimSuffix(errString, ": ")
```

Reply:
nah that's not the suffix the actual error cause typically follows the last `: `